### ISSUE TITLE
LINK : fatal error LNK1189: library limit of 65535 objects exceeded 

### ISSUE THREAD

**Issue by LaurentBerger** [facebookarchive/caffe2#2294]
Windows 10 64 bits
VS 2017
cmake 3.10.2

Using #2029 and #1671
linking caffe2 in debug I have got this errors : 

```
1>------ Build started: Project: caffe2, Configuration: Debug x64 ------
1>LINK : G:\Lib\build\caffe2\bin\Debug\caffe2d.dll not found or not built by the last incremental link; performing full link
1>LINK : fatal error LNK1189: library limit of 65535 objects exceeded
1>Done building project "caffe2.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```
How to solve it?

**Comment by pjh5** [comment]
@LaurentBerger sorry that our Windows coverage isn't so great right now. I think solving this problem will take a lot of effort on our part to narrow down what symbols we want to export. Until then, I think the only workarounds are using release mode or (possibly) linking against less libraries.

**Comment by LaurentBerger** [comment]
I have found a way to solve this problem : use option /Ob2 (see this [issue ](https://github.com/tensorflow/tensorflow/issues/10867#issuecomment-310163123))

Unfortunately I built caffe2 without GPU support and I have got two links errors. I have to disable [those lines
](https://github.com/caffe2/caffe2/blob/master/caffe2/image/image_input_op.h#L1220-L1229) like in this [issue ](https://github.com/caffe2/caffe2/issues/1673#issuecomment-354473512)
Now result is : 
```
1>   Creating library G:/Lib/build/caffe2/lib/Debug/caffe2d.lib and object G:/Lib/build/caffe2/lib/Debug/caffe2d.exp
1>caffe2.vcxproj -> G:\Lib\build\caffe2\bin\Debug\caffe2d.dll
1>Done building project "caffe2.vcxproj".

```

**Comment by cbecker** [comment]
Is this still working with the current master at https://github.com/pytorch/pytorch ?
I tried different options and I'm still hitting the 65535 symbol limit.

**Comment by pjh5** [comment]
@cbecker where you just trying to build Caffe2 or also building Pytorch? Which optional libraries were you including?

**Comment by cbecker** [comment]
@pjh5 I am building Caffe2 from the pytorch repo https://github.com/pytorch/pytorch

I am using the options:
```
USE_OPENCV=OFF
USE_OPENMP=OFF
BUILD_ONNX_PYTHON=OFF
BUILD_CUSTOM_PROTOBUF=ON
BUILD_TEST=OFF
USE_METAL=OFF
USE_GFLAGS=OFF
USE_GLOG=OFF
USE_NNPACK=OFF
USE_NCCL=OFF
CUDA_USE_STATIC_RUNTIME=OFF
CAFFE2_STATIC_LINK_CUDA=OFF
USE_OBSERVERS=OFF
```
And MSVC2017.

For CPU, no CUDA. Do you have any suggestions? I have the impression that it will be almost impossible without splitting the library or manually specifying the symbols to export.

**Comment by pjh5** [comment]
Can you also set BUILD_ATEN and BUILD_BINARY off too? And in Release mode? And the option /Ob2 is being used?

We will have to cut back on the symbols we're exporting eventually, as we really shouldn't run into this limit no matter what libraries we choose to build with.

**Comment by cbecker** [comment]
I'll try that tomorrow (I don't have a Windows machine where I am now) with BUILD_ATEN and BUILD_BINARY off.

I was using /Ob2 and Release, but not sure about these two, are their default values OFF? then they were off.

**Comment by cbecker** [comment]
@pjh5 I tried with BUILD_BINARY and BUILD_ATEN OFF with master from today, and the problem is still there. Is this an issue that will be fixed soon? I am wondering how it's working on CI, as here we cannot compile it with MSVC2017. Maybe because CI runs w/cuda, and we're trying to compile the CPU version here?

**Comment by peterjc123** [comment]
@pjh5 It also occurs when I tried to build pytorch master and the error was thrown in building caffe2 in a Release build with MKL, MAGMA and CUDA 8.0. I think we should prioritize this issue, otherwise, the PyTorch 1.0 can't be built with CUDA 8 on Windows.

**Comment by cbecker** [comment]
I am glad to contribute to this, as I believe it's a big blocker for anyone wanting to use caffe2 on Windows.

I am thinking what would be the best way to go around it. My suggestion is to disable exporting all symbols, not only for windows but also for linux and mac. Then fixing this can be done in linux first, where it's much faster to compile and debug, to include the needed exports. When I tried to do it directly on Windows it was such a pain, because compilation takes much longer in my experience. What's your take?

**Comment by peterjc123** [comment]
Something like sccache and ninja can help to accelerate the build on Windows. 
```CMD
REM Install ninja and sccache
pip install ninja
curl -k https://s3.amazonaws.com/ossci-windows/sccache.exe --output %CD%\\tmp_bin\\sccache.exe
set "PATH=%CD%\\tmp_bin;%PATH%"

REM Specify CMAKE_GENERATOR and C/C++ compiler
set CMAKE_GENERATOR=Ninja
set CC=sccache cl
set CXX=sccache cl
```

**Comment by cbecker** [comment]
Thanks @peterjc123 , that will help.
Nonetheless, wouldn't it then make sense to limit symbol exporting in all platforms, to be consistent and to easy development?

**Comment by peterjc123** [comment]
@cbecker Yes, that will be easier. Please try at https://github.com/pytorch/pytorch with other CIs off by deleting some configurations in `pytorch/pytorch/.jenkins/pytorch/enabled-configs.txt`.
